File: net-event-bus/examples/failover.ts
// The sidecar you no longer run: call a service, not a host.
//
// Two providers serve the same service name. A caller addresses the **service**
// — never a node id — and when the provider that answered first goes away, the
// next call lands on the survivor, with a bounded retry absorbing the transient
// while the dead provider is still in the roster.
//
// Rust has a retry helper (`call_service_typed_with_retry`) that does this for
// you; the SDK has no equivalent, so the loop below is written by hand. That is
// the only difference between this file and its Rust sibling.
//
// Run (from `net/crates/net/sdk-ts`, where `tsconfig.skill-example.json`
// includes it):
//
//   npx tsc --noEmit -p tsconfig.skill-example.json
//
// Expected final line: `RESULT ok providers=2 moved=1 served=2`

import { MeshNode } from '@net-mesh/sdk';
import type { TypedMeshRpc } from '@net-mesh/core/mesh_rpc';

/** 64 hex characters = 32 bytes. Every node in a mesh shares it. */
const PSK = '42'.repeat(32);

/** The roster still lists a dead provider until the fold converges. */
const RETRY_ATTEMPTS = 6;
const RETRY_INTERVAL_MS = 250;
const CALL_DEADLINE_MS = 500;

const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

interface Work {
  units: number;
}
interface Answer {
  served_by: string;
  units: number;
}

async function build(seed: number): Promise<MeshNode> {
  return MeshNode.create({
    bindAddr: '127.0.0.1:0',
    psk: PSK,
    identitySeed: Buffer.alloc(32, seed),
    heartbeatIntervalMs: 200,
  });
}

async function handshake(responder: MeshNode, initiator: MeshNode): Promise<void> {
  const addr = responder.localAddr();
  const pub = responder.publicKey();
  const responderId = responder.nodeId();
  await Promise.all([
    responder.accept(initiator.nodeId()),
    (async () => {
      await sleep(50);
      await initiator.connect(addr, pub, responderId);
    })(),
  ]);
}

async function main(): Promise<void> {
  const caller = await build(0xc1);
  const providerOne = await build(0xc2);
  const providerTwo = await build(0xc3);

  await handshake(caller, providerOne);
  await handshake(caller, providerTwo);
  await caller.start();
  await providerOne.start();
  await providerTwo.start();

  // Two servers, one service name. `serve` advertises the service for each,
  // which is the whole registration.
  const oneId = `0x${providerOne.nodeId().toString(16)}`;
  const twoId = `0x${providerTwo.nodeId().toString(16)}`;
  const rpcOne = providerOne.rpc();
  const rpcTwo = providerTwo.rpc();
  const callerRpc = caller.rpc();

  const serveOne = rpcOne.serve<Work, Answer>('work', (req) => ({
    served_by: oneId,
    units: req.units,
  }));
  const serveTwo = rpcTwo.serve<Work, Answer>('work', (req) => ({
    served_by: twoId,
    units: req.units,
  }));

  // The caller discovers the service by name. It learns *who can serve it*,
  // never whom to prefer — that is what makes the next part work.
  const deadline = Date.now() + 5_000;
  while (Date.now() < deadline) {
    if (callerRpc.findServiceNodes('work').length === 2) break;
    await sleep(25);
  }
  const providers = callerRpc.findServiceNodes('work').length;
  console.log(`providers advertising \`work\`: ${providers}`);

  // Call the service, not a host.
  const first = await callerRpc.callService<Work, Answer>(
    'work',
    { units: 2 },
    { deadlineMs: 1_000 },
  );
  console.log(`first call served by:  ${first.served_by}`);

  // Take that provider out of the mesh entirely — the hard version of a
  // deploy: not a drain, a death. Its RPC handles must be closed first: an
  // outstanding handle makes shutdown fail with "outstanding references exist".
  const doomedIsOne = first.served_by === oneId;
  const doomed = doomedIsOne ? providerOne : providerTwo;
  const doomedServe = doomedIsOne ? serveOne : serveTwo;
  const survivorServe = doomedIsOne ? serveTwo : serveOne;
  const doomedRpc = doomedIsOne ? rpcOne : rpcTwo;
  const survivorRpc = doomedIsOne ? rpcTwo : rpcOne;

  await doomedServe.close();
  await doomedRpc.raw.close();
  await doomed.shutdown();
  console.log(`took ${first.served_by} out`);

  let second: Answer | null = null;
  let lastError: unknown = null;
  for (let attempt = 0; attempt < RETRY_ATTEMPTS && second === null; attempt += 1) {
    try {
      second = await callerRpc.callService<Work, Answer>(
        'work',
        { units: 5 },
        { deadlineMs: CALL_DEADLINE_MS },
      );
    } catch (error) {
      lastError = error;
      await sleep(RETRY_INTERVAL_MS);
    }
  }
  if (second === null) throw lastError ?? new Error('no provider answered after the death');
  console.log(`after the death served by: ${second.served_by}`);

  const moved = second.served_by !== first.served_by ? 1 : 0;
  const served = (first.units === 2 ? 1 : 0) + (second.units === 5 ? 1 : 0);
  console.log(`RESULT ok providers=${providers} moved=${moved} served=${served}`);

  // The survivor's handles, then the caller's.
  await survivorServe.close();
  await survivorRpc.raw.close();
  await callerRpc.raw.close();
  await caller.shutdown();
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});

File: net-event-bus/examples/failover.py
"""The sidecar you no longer run: call a service, not a host.

Two providers serve the same service name. A caller addresses the **service**
— never a node id — and when the provider that answered first goes away, the
next call lands on the survivor, with a bounded retry loop absorbing the
transient while the dead provider is still in the roster.

This is the service-mesh shape collapsed into the bus: no sidecar to
configure, no separate load balancer, no certificate rotation, and no
endpoint list to keep in sync. The roster *is* the capability fold.

Run:

    python failover.py

Expected final line: ``RESULT ok providers=2 moved=1 served=2``
"""

from __future__ import annotations

import threading
import time
from collections.abc import Callable
from typing import Any

from net import NetMesh
from net.mesh_rpc import TypedMeshRpc

# 64 hex characters = 32 bytes. Every node in a mesh shares it.
PSK = "42" * 32

# The service name both providers advertise. The caller never learns a node id.
SERVICE = "work"

# The roster keeps listing the dead provider until the capability fold
# converges, so the first post-death attempt may be spent on the corpse.
# Rust has ``call_service_typed_with_retry``; this binding does not, so the
# loop is written by hand: up to 6 attempts, ~250 ms apart, each bounded by a
# short call deadline. The deadline is the half that matters — a call to a
# corpse does not fail, it waits, so a generous one turns six quick retries
# into half a minute of nothing happening.
RETRY_ATTEMPTS = 6
RETRY_INTERVAL_S = 0.25
CALL_DEADLINE_MS = 500


def build(seed: int) -> NetMesh:
    return NetMesh(
        "127.0.0.1:0",
        PSK,
        identity_seed=bytes([seed]) * 32,
        heartbeat_interval_ms=200,
    )


def handshake(responder: NetMesh, initiator: NetMesh) -> None:
    """One side connects, the other accepts; both calls block."""
    errors: list[BaseException] = []

    def accept() -> None:
        try:
            responder.accept(initiator.node_id)
        except BaseException as error:  # noqa: BLE001 - surfaced below
            errors.append(error)

    thread = threading.Thread(target=accept, daemon=True)
    thread.start()
    time.sleep(0.05)
    initiator.connect(responder.local_addr, responder.public_key, responder.node_id)
    thread.join(timeout=5.0)
    if thread.is_alive():
        raise RuntimeError("handshake timed out")
    if errors:
        raise errors[0]


def handler_for(provider: NetMesh) -> Callable[[dict[str, Any]], dict[str, Any]]:
    """Build a handler that echoes its own provider's node id.

    The caller never asked for a provider by id, so ``served_by`` is the only
    way to prove which one ran the work.
    """
    provider_id = provider.node_id

    def run(req: dict[str, Any]) -> dict[str, Any]:
        return {"served_by": provider_id, "units": req["units"]}

    return run


def main() -> None:
    caller = build(0xC1)
    provider_one = build(0xC2)
    provider_two = build(0xC3)

    # Each provider connects to the caller: the caller is the responder.
    handshake(caller, provider_one)
    handshake(caller, provider_two)
    caller.start()
    provider_one.start()
    provider_two.start()

    caller_rpc = TypedMeshRpc.from_mesh(caller)
    one_rpc = TypedMeshRpc.from_mesh(provider_one)
    two_rpc = TypedMeshRpc.from_mesh(provider_two)

    one_id = provider_one.node_id
    two_id = provider_two.node_id

    try:
        # Two servers, one service name. ``serve`` advertises ``nrpc:work`` for
        # each, which is the whole registration.
        with one_rpc.serve(SERVICE, handler_for(provider_one)), two_rpc.serve(
            SERVICE, handler_for(provider_two)
        ):
            # The caller discovers the service by name. It learns *who can
            # serve it*, never whom to prefer — that is what makes the next
            # part work.
            deadline = time.monotonic() + 5.0
            while time.monotonic() < deadline:
                if len(caller_rpc.find_service_nodes(SERVICE)) == 2:
                    break
                time.sleep(0.025)
            providers = len(caller_rpc.find_service_nodes(SERVICE))
            print(f"providers advertising `{SERVICE}`: {providers}")

            # Call the service, not a host.
            first = caller_rpc.call_service(
                SERVICE, {"units": 2}, opts={"deadline_ms": 1000}
            )
            print(f"first call served by:  {first['served_by']:#x}")

            # Take that provider out of the mesh entirely — the hard version
            # of a deploy: not a drain, a death.
            if first["served_by"] == one_id:
                provider_one.shutdown()
            else:
                provider_two.shutdown()
            print(f"took {first['served_by']:#x} out")

            # The roster still lists the dead provider until the capability
            # fold converges, so the first attempt may be spent on it. Retry
            # by hand to make that a transient rather than a caller-visible
            # failure.
            second: dict[str, Any] | None = None
            last_error: BaseException | None = None
            for _ in range(RETRY_ATTEMPTS):
                try:
                    second = caller_rpc.call_service(
                        SERVICE, {"units": 5}, opts={"deadline_ms": CALL_DEADLINE_MS}
                    )
                    break
                except Exception as error:  # noqa: BLE001 - the corpse is retried
                    last_error = error
                    time.sleep(RETRY_INTERVAL_S)
            if second is None:
                raise (
                    last_error
                    if last_error is not None
                    else RuntimeError("no answer after retries")
                )
            print(f"after the death served by: {second['served_by']:#x}")

            moved = int(second["served_by"] != first["served_by"])
            served = int(first["units"] == 2) + int(second["units"] == 5)
            print(f"RESULT ok providers={providers} moved={moved} served={served}")
    finally:
        caller.shutdown()
        provider_one.shutdown()
        provider_two.shutdown()


if __name__ == "__main__":
    main()